File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {

  private final int speed;
  private final int batteryDrain;
  private int totalBattery = 100;
  private int totalDistanceDriven = 0;

  NeedForSpeed(int speed, int batteryDrain) {
    this.batteryDrain = batteryDrain;
    this.speed = speed;
  }

  public boolean batteryDrained() {
    return totalBattery == 0;
  }

  public int distanceDriven() {
    return totalDistanceDriven;
  }

  public void drive() {
    if (totalBattery != 0) {
      this.totalBattery -= this.batteryDrain;
      this.totalDistanceDriven += this.speed;
    }
  }

  public static NeedForSpeed nitro() {
    return new NeedForSpeed(50, 4);
  }
}

class RaceTrack {

  private final int distance;

  RaceTrack(int distance) {
    this.distance = distance;
  }

  public boolean tryFinishTrack(NeedForSpeed car) {
    while (!car.batteryDrained()) {
      car.drive();
    }

    return car.distanceDriven() >= distance;
  }
}

File: go/need-for-speed/need_for_speed.go
package speed

type Car struct {
	battery      int
	batteryDrain int
	speed        int
	distance     int
}

type Track struct {
	distance int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	return Car{
		battery:      100,
		distance:     0,
		speed:        speed,
		batteryDrain: batteryDrain,
	}
}

// NewTrack creates a new track
func NewTrack(distance int) Track {
	return Track{distance: distance}
}

// Drive drives the car one time. If there is not enough battery to drive one more time,
// the car will not move.
func Drive(car Car) Car {
	if car.batteryDrain <= car.battery {
		car.battery -= car.batteryDrain
		car.distance += car.speed
	}
	return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
	if car.batteryDrain < car.battery {
		return (car.battery - (track.distance/car.speed)*car.batteryDrain) >= 0
	}
	return false
}
